---
title: "RNAseq123"
output: html_notebook
---

Goal: to recap RNA-seq analysis.



code: https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.R
article: https://f1000research.com/articles/5-1408/v3


about the data:

Workflow is from Sheridan et al. (2015).
this?: https://pubmed.ncbi.nlm.nih.gov/26080807/
or this?: https://pubmed.ncbi.nlm.nih.gov/25879659/

3 cell populations in triplicate:
  - basal
  - LP: luminal progenitor
  - ML: mature luminal
data:
  - 100 base-pair single-end reads, aquired by Illumina HiSeq 2000.
  - count data: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63310
  - mouse reference genome (mm10)
method:
  - align reads to mm10 reference genome using the R based pipeline available in the Rsubread package (specifically the align function6 followed by featureCounts7 for gene-level summarisation based on the in-built mm10 RefSeq-based annotation).


In this instance, reads were aligned to the mouse reference genome (mm10) using the R based pipeline available in the Rsubread package
Count data for these samples can be downloaded from the Gene Expression Omnibus (GEO) http://www.ncbi.nlm.nih.gov/geo/ using GEO Series accession number GSE63310. Further information on experimental design and sample preparation is also available from GEO under this accession number.

the study:
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63310

Purpose: The aim of this study is to determine the absolute and relative expression levels of mRNA transcripts across multiple flow cytometrically sorted epithelial cell types including freshly isolated CD24+CD29hi mammary stem cell-enriched basal cells (MaSC/basal), CD24+CD29loCD61+ luminal progenitor-enriched (LP) and the CD24+CD29loCD61- mature luminal-enriched (ML) cell populations. Additionally, a comparison between these primary cell types and cultured MaSC/Basal-derived mammosphere cells (mammosphere) and the CommaD-βGeo (CommaDβ) cell line was performed.

Methods: Total RNA was extracted and purified from sorted luminal or basal populations from the mammary glands of female virgin 8- to 10-week-old FVB/N mice (3 independent samples per population), MaSC/Basal cells cultured for 1 week under mammosphere conditions and CommaDβ cells grown under maintenance conditions (Deugnier et al. 2006). Total RNA (100 ng) was used to generate sequencing libraries for whole transcriptome analysis following Illumina’s TruSeq RNA v2 sample preparation protocol. Completed libraries were sequenced on HiSeq 2000 with TruSeq SBS Kit v3- HS reagents (Illumina) as 100 bp single-end reads at the Australian Genome Research Facility (AGRF), Melbourne. Approximately 30 million 100 bp single-end reads were obtained for each sample. Reads were aligned to the mouse reference genome mm10 and mapped to known genomic features at the gene level using the Rsubread package (version 1.14.1) (Liao et al. 2013). Single reads were then summarized into gene-level counts using FeatureCounts (Liao et al. 2014).

mammosphere: sphere of cells that is formed from a single mammalian stem cell.

CD: antibodies.

luminal progenitor-enriched: a cell that can differentiate into different cell types cells from the inside of an organ that stores or transports stuff.

cytromery:
https://en.wikipedia.org/wiki/Cytometry
Cytometry is the measurement of number and characteristics of cells. Variables that can be measured by cytometric methods include cell size, cell count, cell morphology (shape and structure), cell cycle phase, DNA content, and the existence or absence of specific proteins on the cell surface or in the cytoplasm.[

CD24:
https://pubmed.ncbi.nlm.nih.gov/36013184/
Cluster of differentiation 24 (CD24) is a small, highly glycosylated cell adhesion protein that is normally expressed by immune as well as epithelial, neural, and muscle cells. Tumor CD24 expression has been linked with alterations in several oncogenic signaling pathways. In addition, the CD24/Siglec-10 interaction has been implicated in tumor immune evasion, inhibiting macrophage-mediated phagocytosis as well as natural killer (NK) cell cytotoxicity.


CD29: 
https://en.wikipedia.org/wiki/Integrin_beta_1
Integrin beta-1 (ITGB1), also known as CD29, is a cell surface receptor that in humans is encoded by the ITGB1 gene.[5] This integrin associates with integrin alpha 1 and integrin alpha 2 to form integrin complexes which function as collagen receptors. It also forms dimers with integrin alpha 3 to form integrin receptors for netrin 1 and reelin. These and other integrin beta 1 complexes have been historically known as very late activation (VLA) antigens.

Integrin beta 1 is expressed as at least four different isoforms. In cardiac muscle and skeletal muscle, the integrin beta-1D isoform is specifically expressed, and localizes to costameres, where it aids in the lateral force transmission from the Z-discs to the extracellular matrix. Abnormal levels of integrin beta-1D have been found in limb girdle muscular dystrophy and polyneuropathy.
  
CD61:
https://en.wikipedia.org/wiki/Integrin_beta_3
Integrin beta-3 (β3) or CD61 is a protein that in humans is encoded by the ITGB3 gene.[5] CD61 is a cluster of differentiation found on thrombocytes.[6]
The ITGB3 protein product is the integrin beta chain beta 3. Integrins are integral cell-surface proteins composed of an alpha chain and a beta chain. A given chain may combine with multiple partners resulting in different integrins. Integrin beta 3 is found along with the alpha IIb chain in platelets. Integrins are known to participate in cell adhesion as well as cell-surface-mediated signaling.[7]
Defectively expressed β3 integrin subunit has been correlated with presence of endometriosis, and has been suggested as a putative marker of this condition.[8]
CD61 has been shown to interact with PTK2,[9][10] ITGB3BP,[11][12] TLN1[13][14] and CIB1.[15]

# Let's start the excercise!

```{r}
library(limma)
library(Glimma)
library(edgeR)
library(tidyr) # for reshaping data
library(dplyr) # other data manilulation
library(Mus.musculus)
library(ggplot2)
library(grid)
ibrary(gridExtra)
```
# Data packaging

## Reading in count data

Download and extract count data.
Important: there are two files that are not used: 
  - "GSM1545537_mo906111-1_m09611-2.txt"
  - "GSM1545543_JMS9-CDBG.txt"

```{r}
# # download
# url <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE63310&format=file"
# utils::download.file(url, destfile="data/GSE63310_RAW.tar", mode="wb")
# utils::untar("data/GSE63310_RAW.tar", exdir = "data/")
# 
# # extract
# # files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", "GSM1545538_purep53.txt",
# #   "GSM1545539_JMS8-2.txt", "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt", 
# #   "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", "GSM1545545_JMS9-P8c.txt",
# #   "GSM1545537_mo906111-1_m09611-2.txt", "GSM1545543_JMS9-CDBG.txt")
# 
# files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", "GSM1545538_purep53.txt",
#   "GSM1545539_JMS8-2.txt", "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt",
#   "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", "GSM1545545_JMS9-P8c.txt")
# 
# for(i in paste("data/", files, ".gz", sep=""))
#   R.utils::gunzip(i, overwrite=TRUE)
```

Look at first file in array.
I put in a duplicate of the files variable, in case the files were already downloaded and extracted.
```{r}
files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", "GSM1545538_purep53.txt",
  "GSM1545539_JMS8-2.txt", "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt",
  "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", "GSM1545545_JMS9-P8c.txt")

read.delim(paste("data/",files[1],sep=""), nrow=5)
```

We can lead all the count files in 1 go using edgeR!

I made a mistake and didn't tell readDGE to select specific columns.
This resulted in equal lib siszes.
if readDGE takes in a file, it just takes the first two columns.
So it only needs files and libsizes of the files.

If you look at the help page of readDGE, by default it uses file names as the label of a sample.
Following the help file, it needs 2 data columns: gene names, and counts.
So the two columns needed are EntrezID and Count.
if it takes geneLength

3634
wikipedia
https://en.wikipedia.org/wiki/Library_(biology)
library = library is a collection of DNA fragments that is stored and propagated in a population of micro-organisms through the process of molecular cloning

question: is library size the amount of sequences, or the total amount of nucleotides?
answer: the amount of molecules (as in: read sequences)
https://en.wikipedia.org/wiki/RNA-Seq
"Sequencing depth is sometimes referred to as library size, the number of intermediary cDNA molecules in the experiment."
Note: Sequencing depth == library size (2 words for the same thing)


```{r}
x <- readDGE(files=files, path="data/", columns=c(1,3))
x
```

## Organsing sample information

For simplicity, we remove the GEO sample IDs (GSM*) from the column names of our DGEList-object x.
The example states that we need to state the length of the string, but this is not needed.
```{r}
samplenames <- substring(colnames(x), 12)
samplenames
```

```{r}
x$samples
```
we give them the proper names.
We also add lane information.
```{r}
colnames(x) <- samplenames
group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP"))
x$samples$group <- group
lane <- as.factor(rep(c("L004","L006","L008"), c(3,4,2)))
x$samples$lane <- lane
x$samples
```
## Organising gene annotations
We are ging to store gene-level information about the counts into the DGElist-object.
the rownames of our DGElist-object contain Entrez gene-id's.
We will use those to collect data from biomaRt about the symbol of the gene and the chromosome on which the gene is on.
```{r}
library(Mus.musculus)
geneid <- rownames(x)
genes <- select(Mus.musculus, keys=geneid, columns=c("SYMBOL","TXCHROM"), keytype="ENTREZID")
dim(genes)
head(genes)
```

There are duplicate ID's, let's take a look.
```{r}
arrange(genes[duplicated(genes$ENTREZID),],ENTREZID)
```
let's do a check
```{r}
genes[genes$ENTREZID=="100042555",]
```
As we can see they have different instances.
For simplicity, we are going to take the first occurrence.
Not the best method, but it doesn't matter for learning purposes.
```{r}
genes <- genes[!duplicated(genes$ENTREZID),]
x$genes <- genes
```

Keep in mind that x$genes is ordered the same as the other data, because we only took the first onthology by using duplicated().
if this is not the case, you can use the matchfunction to order correctly.

what we do:
1) Data pre-processing:
  - transforming from the raw-scale
  - Removing genes that are lowly expressed
  - Normalising gene expression distributions

# Data pre-processing

## transforming from the raw-scale

Ways to deal with library size differences:
- counts per million (cpm)
- log2-cpm
- reads per kilobase of transcript per million (RPKM)
- fragments per kilobase of transcripts per million (FPKM)

edgeR can calculate cpm and RPKM, we will use cpm.
It just takes the count, divides it by the library size and multiplies the answer with 10^6.
```{r}
cpm <- cpm(x)
lcpm <- cpm(x, log=TRUE)
head(cpm)
```
I'm very curious if it takes the library size into account automatically.
let's test it.
```{r}
backup = x$samples[1,"lib.size"]
```

```{r}
x$samples[1,"lib.size"] = 32132132
test <- cpm(x)
head(test)
head(cpm)
x$samples[1,"lib.size"] = backup
```
It takes library size into account!
Not much of a suprise, but good to see confirmation.
```{r}
x$samples
```
The lib size of 10_6_5_11 is: 32.863.052.
So 1 cpm for a gene is equal to 32.8 counted reads.

log-cpm are used for exploration of the data.
log transform scales down extreme values, while leaving smaller values more the same (relatively speaking).
lcpm() uses 2log by default, wich is handy.
lcpm() doesn't take the 2log of 0, which is undefined, which is nice.

In the example the look at the mean and median of the library sizes as cpm,so we divide by 10^6.
```{r}
L <- mean(x$samples$lib.size) / 10^6
M <- median(x$samples$lib.size) * 1e-6
c(L, M)
```

```{r}
summary(lcpm)
```
Keep in mind that negative lcpm are just cpm, smaller than 1.

note: voom (from limma) uses log-cpm, although voom recomputes its own log-CPM values internally with a smaller prior count.

## Removing genes that are lowly expressed

x$counts==0 : get a true for every gene that has no counts regardless of sample.
rowSums(x$counts==0) : sums the rows (it sums the TRUE values (so the 1's)).
rowSums(x$counts==0)==9 :  gives a true for 9 trues in a row, so a true for every column/sample. In other words there are no counts for this gene in any of the samples.
table(rowSums(x$counts==0)==9) : count rows without any value and with a value.
```{r}
table(rowSums(x$counts==0)==9)
```
```{r}
table(rowSums(x$counts==0)==9)["TRUE"] / (table(rowSums(x$counts==0)==9)["FALSE"] + table(rowSums(x$counts==0)==9)["TRUE"]) * 100
```
19% of the genes have no counts in all samples.

The figure in the example states that there a re a low of low expressed genes.
I'm going to try plotting that graph from scratch, no hand-holding.
The local minimum is at about 0, but i'm too lazy to figure that one out, so...
```{r}
melted <- data.frame(lcpm) %>% gather()
names(melted) = c("sample","lcpm")
plot.bf <- ggplot(data=melted, aes(x=lcpm, color=sample, )) + 
  geom_density() + 
  labs(title="Density plot of lcpm before filtering") +
  geom_vline(xintercept=0.2)
plot_bf
```
I'ts a nicer plot with less code, this is hilarious.

Automatic filtering by expression level.
Group takes in the name of the group column from the samples datasheet.
I'm not sure about the third argument.
It's a DGEList object, with an label var, column var, and the third must be other options??
A bit weird way to do it like that, oh well....

IMPORTANT NOTE: - we filter on the original counts, not the logcount!
                - keep in mind that x is now filtered.

```{r}
keep.expr <- filterByExpr(x, group=group)
x.no_low_expr <- x[keep.expr,,keep.lib.sizes=FALSE]
dim(x.no_low_expr)
```
By default it keeps genes that have at least 10 read counts.
In reality the function calculates cpm and calculates a cut-off that correspondents to 10 read counts.
There are 3 replicates per cell group. 
NOTE: An biological interesting gene should be expressed in at least all replicates.

The cutoff used depends on the sequencing depth and the experimental design.
IMPORTANT:
A larger library size gives a better resolution at lower expression levels, so then we can take a lower cpm.
A smaller libsize gives worse resolution at lower expressed levels, so then we must take a higer cpm.

we now have 16624 genes, before we had 27208.
```{r}
(16624 - 27208) / 27208 * 100
```
So a 38,9% reduction of genes.

Note that subsetting the entire DGEList-object removes both the counts and the associated gene information for the filtered genes. The filtered DGEList-object keeps the gene information and the counts for the retained genes correctly associated.

So always use filterByExpr().

```{r}
x
```
```{r}
x.no_low_expr
```

Now we make the graph again, but first calculate lcpm again.
```{r}
# .nl = no low expressed genes
cpm.nl <- cpm(x.no_low_expr)
lcpm.nl <- cpm(x.no_low_expr, log=TRUE)

melted.nl <- data.frame(lcpm.nl) %>% gather()
names(melted.nl) = c("sample","lcpm")

plot.nl <- ggplot(data=melted.nl, aes(x=lcpm, color=sample, )) + 
  geom_density() + 
  labs(title="Density plot of lcpm before filtering") +
  geom_vline(xintercept=0.2)
plot.nl
```
Let's put them side by side for the hell of it.
I need to find a better package for this....
I used to have a good one, but can't remember it's name..
```{r}
grid.arrange(plot.nl, plot.bf, ncol=2)
```
## Normalising gene expression distributions
It can be that an sample has an overal higher expression than another one.
This can occur due to external factors, which gives a skewed result for the samples.
"It is assumed that all samples should have a similar range and distribution of expression values."
so we arew going to normalize the gene expression <b>distribution</b> across samples.

You can see in the previous plots that the distributions are close to being the same (all the lines moslty overlap).
So it isn't really needed, but we do it anyway, for learning sake.
The normalisation we will use is trimmed mean of M-values (TMM).

Oeh! This is handy: an explenation of all gene expression units:
https://www.reneshbedre.com/blog/expression_units.html#:~:text=TMM%20%28Trimmed%20Mean%20of%20M-values%29%20Permalink.%20TMM%20is,most%20of%20the%20genes%20are%20not%20differentially%20expressed.

- TMM is a between-sample normalization method in contrast to within-sample normalization methods (RPM, RPKM/FPKM, or TPM)
- TMM normalization method assumes that most of the genes are not differentially expressed
- TMM normalize the total RNA output among the samples and does not consider gene length or library size for normalization (NOTE: so great with cpm(?)).
- TMM considers sample RNA population and effective in normalization of samples with diverse RNA repertoires (e.g. samples from different tissues). TMM will be good choice to remove the batch effects while comparing the samples from different tissues or genotypes or in cases where RNA population would be significantly different among the samples.
- TMM is implemented in edgeR and performs better for between-samples comparisons
- edgeR does not consider gene length for normalization as it assumes that the gene length would be constant between the samples
- to calculate:
  M = log 2 fold change between two samples
  M = log2(treated sample count / control sample count)
  A = absolute expression count
  A = (log2(treated sample count) + log2(control sample count)) / 2
  double trimm the lower and upper percentages of the data:
    trim M by 30% and a by 5%.
  get weighted mean of M after trimming and calculate normalisation factor.
  Robinson et al., 2010.: https://pubmed.ncbi.nlm.nih.gov/19910308/

```{r}
# puts normalisation factors in x$samples$norm.factors
x <- calcNormFactors(x.no_low_expr, method = "TMM")
x$samples$norm.factors
```

The effects of normalisation are a bit small, so we.....

TO BE CONTINUED!!
https://f1000research.com/articles/5-1408/v3#f1
